# **My Tokenizer**

Beyza Balota - 31232

20.10.2024

## 1. Tokenization

In [1]:
import re

def my_tokenizer(corpus_raw):
    '''
    type corpus_raw: string
    param corpus_raw: The raw output of the corpus to be tokenized
    rtype: list
    return: a list of tokens extracted from the corpus_raw
    '''

    # Step 1: Remove content inside square brackets (e.g., [t], [+2])
    corpus_raw = re.sub(r"\[.*?\]", "", corpus_raw)

    # Step 2: Remove unnecessary symbols like ##
    corpus_raw = re.sub(r"##", "", corpus_raw)

    # Step 3: Handle punctuation and words
    # Find all sequences of alphanumeric characters, including hyphenated words, or separate punctuation marks
    token_list = re.findall(r"\w+(?:-\w+)*|[.,!?;()\"']", corpus_raw)

    # Step 4: Handle common contractions (e.g., don't -> do not)
    token_list = [re.sub(r"n't", " not", token) for token in token_list]
    token_list = [re.sub(r"'re", " are", token) for token in token_list]
    token_list = [re.sub(r"'s", "'is", token) for token in token_list]  
    token_list = [re.sub(r"'ve", " have", token) for token in token_list]
    token_list = [re.sub(r"'ll", " will", token) for token in token_list]

    # Step 5: Convert tokens to lowercase
    token_list = [token.lower() for token in token_list]

    # Step 6: Filter out any empty strings or stray characters
    token_list = [token for token in token_list if token.strip()]

    return token_list

## 2. Choosing and Dowloading the Corpus & Calling my Tokenizer Function

In [2]:
#main code

#using nltk to download my selected corpus
import nltk 

corpus_name = 'product_reviews_2'

#download the corpus and import it.
nltk.download('product_reviews_2')
from nltk.corpus import product_reviews_2

#get the raw text output of the corpus to the corpus_raw variable
corpus_raw = product_reviews_2.raw()

#call tokenizer method
my_tokenized_list = my_tokenizer(corpus_raw)


[nltk_data] Downloading package product_reviews_2 to
[nltk_data]     /Users/beyzabalota/nltk_data...
[nltk_data]   Package product_reviews_2 is already up-to-date!


## 3. Displaying Raw and Tokenized Versions of the Corpus 
(decoding purposes)

In [3]:
#displaying for decoding purposes
from IPython.display import display

# Output the first 200 characters of the raw text
display("Original raw text:", corpus_raw[:200])

# Output the first 20 tokens from the tokenized list
display("Tokenized output:", my_tokenized_list[:20])


'Original raw text:'

"[t]\nSD500[+2]##We really enjoyed shooting with the Canon PowerShot SD500. \ndesign[+2]##It has an exterior design that combines form and function more elegantly than any point-and-shoot we've ever test"

'Tokenized output:'

['sd500we',
 'really',
 'enjoyed',
 'shooting',
 'with',
 'the',
 'canon',
 'powershot',
 'sd500',
 '.',
 'designit',
 'has',
 'an',
 'exterior',
 'design',
 'that',
 'combines',
 'form',
 'and',
 'function']

## Please do not touch the code below that will evaluate your tokenizer with the nltk word tokenizer. You will get zero points from evaluation if you do so.

In [4]:
def similarity_score(set_a, set_b):
    '''
    type set_a: set
    param set_a: The first set to be compared
    type set_b: set
    param set_b: The tokens extracted from the corpus_raw
    rtype: float
    return: similarity score with two sets using Jaccard similarity.
    '''

    jaccard_similarity = float(len(set_a.intersection(set_b)) / len(set_a.union(set_b)))

    return jaccard_similarity

In [5]:
from nltk import word_tokenize
nltk.download('punkt')
from nltk import punkt

def evaluation(corpus_raw, token_list):
    '''
    type corpus_raw: string
    param corpus_raw: The raw output of the corpus
    type token_list: list
    param token_list: The tokens extracted from the corpus_raw
    rtype: float
    return: comparison score with the given token list and the nltk tokenizer.
    '''

    #The comparison score only looks at the tokens but not the frequencies of the tokens.
    #we assume case folding is already applied to the token_list
    corpus_raw = corpus_raw.lower()
    nltk_tokens = word_tokenize(corpus_raw, language='english')

    score = similarity_score(set(token_list), set(nltk_tokens))

    return score

[nltk_data] Downloading package punkt to
[nltk_data]     /Users/beyzabalota/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [6]:
#Evaluation

eval_score = evaluation(corpus_raw, my_tokenized_list)

print('The similarity score is {:.2f}'.format(eval_score))


The similarity score is 0.80


## **4. Report**

## Corpus Name and Selection Reason
For this assignment, I selected the product_reviews_2 corpus from the NLTK library. The reason behind this selection was that product reviews generally contain a mix of everyday language, domain-specific terminology, abbreviations, and punctuation. This makes it an interesting case for tokenization, especially as reviews may also include numbers, hyphenated terms, and contractions. I believed that this corpus would be a challenge for me to create a custom tokenizer 

## Design of the Tokenizer
I have designed the tokenizer using the top-down approach. Firstly, I started with downloading the corpus using the NLTK library, then I displayed the first 200 characters to see the behaviour of the corpus, meaning that I could see unnecessary punctions or braces and get rid of them at the earlier stage. After displaying the corpus, I realized that there are many characters that are inside the square brackets and making noise inside the text, so I removed them. Furthermore, I removed some repeated symbols like “##” that were assumed unnecessary for tokenization.  I decided not to separate hyphenated words (e.g., " point-and-shoot") since breaking them might cause losing their original meaning. Later on, I decided to handle alphanumeric characters to ensure that words, product names, and numbers are tokenized correctly, such as "Canon123" or "camera. I also handled punctuation in this part by separating them from the words, which is a common practice in tokenization. The reason behind that separation for me was mainly about achieving more accurate parsing and token matching. After that, I continued with contradiction handling, which I found out after some research that expanding them such as “don’t” to “do not” and “you’re” to “you are” was an excellent approach for maintaining the meaning of the words.  The final step was lowercasing for case insensitivity, which was also mentioned by Dilara Hoca in her e-mail, and removing empty tokens.

## Challenges Faced During Tokenization
During this project, I faced some difficulties. My primary challenge was handling the balance between splitting punctuation and retaining meaning in the text. Product reviews often use punctuation in informal ways, such as ellipses (...) or multiple exclamation marks (!!!). Deciding whether to keep these as individual tokens or merge them with words was tricky. Another challenge was managing hyphenated words and numeric values. As been discussed before, I aimed to keep hyphenated words intact, but there are edge cases where a more advanced algorithm could better decide whether the hyphen is meaningful or not.

## Limitations of the Approach & Possible Improvements
When it comes to the limitation of the approach, the handling of more complex punctuation and symbols can be considered. For example, repeated punctuation marks like ellipses (...) are still treated as three separate tokens. The treatment of hyphenated words is basic; it works well for many cases but could be improved with more context awareness. Additionally, expanding contractions might not always work perfectly, especially with unusual uses or variations in how people speak. The tokenizer also doesn't understand the context, so it can't adjust its behavior based on the specific content (for example, treating specialized terms differently from everyday words). Possible improvements can be made, especially in the case of contextual tokenization using machine learning models. Other than that, the tokenizer could be enhanced to deal with abbreviations and product specific terminology more effectively, such as recognizing that "25mm" or "3D" are some specific meaningful terms that should not be split.

## Why This Approach May Be Better than NLTK
While NLTK's word tokenizer is widely used, it does not handle some contractions and hyphenated words as effectively as the tokenizer I have implemented in my project. My approach makes deliberate efforts to expand contractions, handle possessive cases, and manage specific terms like hyphenated words and numbers more carefully. This results in more contextually accurate tokens that reflect the true meaning of product reviews. However, the evaluation score is based on similarity to NLTK’s tokenizer, which limits the score despite these enhancements. I got 0.80 similarity, which is quite good. Despite this, my approach could be better for future tasks. 
